SpatialGDK/Source/SpatialGDK/Private/SpatialView/CriticalSectionFilter.cpp

```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialView/CriticalSectionFilter.h"
+
+#include "SpatialView/OpList/SplitOpList.h"
+
+namespace SpatialGDK
+{
+void FCriticalSectionFilter::AddOpList(OpList Ops)
+{ // Ensure that we only process closed critical sections.
+	// Scan backwards looking for critical sections ops.
+	for (uint32 i = Ops.Count; i > 0; --i)
+	{
+		Worker_Op& Op = Ops.Ops[i - 1];
+		if (Op.op_type != WORKER_OP_TYPE_CRITICAL_SECTION)
+		{
+			continue;
+		}
+
+		// There can only be one critical section open at a time.
+		// So any previous open critical section must now be closed.
+		for (OpList& OpenCriticalSection : OpenCriticalSectionOps)
+		{
+			ReadyOps.Add(MoveTemp(OpenCriticalSection));
+		}
+		OpenCriticalSectionOps.Empty();
+
+		// If critical section op is opening the section then enqueue any ops before this point and store the open critical section.
+		if (Op.op.critical_section.in_critical_section)
+		{
+			SplitOpListPair SplitOpLists(MoveTemp(Ops), i);
+			ReadyOps.Add(MoveTemp(SplitOpLists.Head));
+			OpenCriticalSectionOps.Add(MoveTemp(SplitOpLists.Tail));
+		}
+		// If critical section op is closing the section then enqueue all ops.
+		else
+		{
+			ReadyOps.Add(MoveTemp(Ops));
+		}
+		return;
```

Comment:
I'm not sure why this wouldn't return earlier than we'd want if there were critical section ops in the op list before a critical section. Can that not happen? And if not can this be documented? Or have I misunderstood this logic

Reply:
(I realise the logic is just a copy paste so I am fairly sure it works, just curious)

Reply:
This iterates backwards through the list.

Reply:
So first critical section op you find tells you everything you need to know.

Reply:
This could probably do with a longer comment at the top explaining it.